### PR TITLE
Add dataset filtering script

### PR DESCRIPTION
**Pull request by edbeeching**
This PR adds a script that generates `k` responses from a model and scores the reward using reward function(s).
The mean reward can be used to filter the dataset of easy and hard problems, which may improve GRPO performance.

Usage:

```shell
python scripts/filter_dataset.py --config recipes/dataset_filtering/config_demo.yaml
```

Example dataset, run on a 1k subset:

[OpenR1-Math-220k-generated](https://huggingface.co/datasets/open-r1/OpenR1-Math-220k-generated)

[OpenR1-Math-220k-generated-filtered](https://huggingface.co/datasets/open-r1/OpenR1-Math-220k-generated-filtered)

### CODE REVIEW

File: recipes/dataset_filtering/config_demo.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+dataset_name: open-r1/OpenR1-Math-220k
+dataset_prompt_column: problem
+system_prompt: "You are a helpful AI Assistant that provides well-reasoned and detailed responses. You first think about the reasoning process as an internal monologue and then provide the user with the answer. Respond in the following format: <think>\n...\n</think>\n<answer>\n...\n</answer>"
+
+# Generation arguments
+max_completion_length: 2048
+num_generations: 4
```
Comment (lewtun) [inline recipes/dataset_filtering/config_demo.yaml:16]: Let's set 8 as a good default
```suggestion
num_generations: 8
```

File: scripts/filter_dataset.py
Comment (lewtun) [inline scripts/filter_dataset.py:1]: To make it explicit, let's call this `compute_pass_rate.py`

File: scripts/filter_dataset.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    
+    dataset = dataset.map(apply_chat_template, fn_kwargs={"tokenizer": tokenizer})
+    llm = LLM(
+        model=model_args.model_name_or_path,
+        revision=model_args.model_revision,
+        trust_remote_code=model_args.trust_remote_code,
+    )
+
+    sampling_params=SamplingParams(
+        temperature=training_args.temperature,
```
Comment (lewtun) [inline scripts/filter_dataset.py:103]: Expose `top_p` here

File: scripts/filter_dataset.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            reward_kwargs = {key: repeat_each_element_k_times(examples[key], training_args.num_generations) for key in keys}
+            output_reward_func = reward_func(prompts=repeated_prompts, completions=reward_completions, **reward_kwargs)
+            # Convert None values to NaN
+            output_reward_func = [reward if reward is not None else torch.nan for reward in output_reward_func]
+
+            rewards_per_func[:, i] = torch.tensor(output_reward_func, dtype=torch.float32)
+            
+        reshaped_rewards = rewards_per_func.view(-1, training_args.num_generations)
+        
+        examples["filter_generations"] = grouped_completions
```
Comment (lewtun) [inline scripts/filter_dataset.py:145]: Let's be explicit here since filter can mean many things:
```suggestion
        examples["pass_rate_generations"] = grouped_completions
```

File: scripts/filter_dataset.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        examples["filter_rewards"] = reshaped_rewards.tolist()
```
Comment (lewtun) [inline scripts/filter_dataset.py:146]:
```suggestion
        examples["pass_rate_rewards"] = reshaped_rewards.tolist()
```

File: scripts/filter_dataset.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
```
Comment (lewtun) [inline scripts/filter_dataset.py:147]:
```suggestion
            example["pass_rate_generator"] = model_args.model_name_or_path
```

File: scripts/filter_dataset.py (hunk 3 above)
Comment (lewtun) [reply scripts/filter_dataset.py:145]: Let's also add a `pass_rate` column, so it's easy to do additional filtering if the original thresholds aren't optimal

File: scripts/filter_dataset.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    else:
+        unfiltered_dataset_name = f"{script_args.dataset_name}-generated"
+    filtered_dataset_name = f"{unfiltered_dataset_name}-filtered"
+    
+    dataset.push_to_hub(unfiltered_dataset_name)
+    
+    def filter_func(example):
+        rewards = example["filter_rewards"]
+        # get the mean of the rewards that are not None
+        mean_reward = torch.nanmean(torch.tensor(rewards, dtype=torch.float32))
```
Comment (lewtun) [inline scripts/filter_dataset.py:164]: This can be done above when we include a `pass_rate` column and then this filter could be a lambda function

File: scripts/filter_dataset.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    dataset = dataset.map(batch_score, batched=True, batch_size=256)
+    
+    
+    if script_args.output_dataset_name is not None:
+        unfiltered_dataset_name = script_args.output_dataset_name
+    else:
+        unfiltered_dataset_name = f"{script_args.dataset_name}-generated"
+    filtered_dataset_name = f"{unfiltered_dataset_name}-filtered"
+    
+    dataset.push_to_hub(unfiltered_dataset_name)
```
Comment (lewtun) [inline scripts/filter_dataset.py:159]: Instead of having a proliferation of datasets, how about we use dataset configs to denote the unfiltered vs filtered subsets, e.g.:

* `unfiltered` = this one
* `filtered_{script_args.low_reward_threshold}-{script_args.high_reward_threshold}`

That way we can do future filterings and have them all in one place.

File: scripts/filter_dataset.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+            
+        return examples
+        
+    dataset = dataset.map(batch_score, batched=True, batch_size=256)
+    
+    
+    if script_args.output_dataset_name is not None:
+        unfiltered_dataset_name = script_args.output_dataset_name
+    else:
+        unfiltered_dataset_name = f"{script_args.dataset_name}-generated"
```
Comment (lewtun) [inline scripts/filter_dataset.py:156]:
```suggestion
        unfiltered_dataset_name = f"{script_args.dataset_name}-pass-rate"
```

File: recipes/dataset_filtering/config_demo.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+temperature: 0.7
+
+# Reward func arguments
+reward_funcs:
+- accuracy
+reward_weights:
+- 1.0
+
+# Filtering arguments. Samples with mean reward outside of low / high will be filtered
+low_reward_threshold: 0.2
```
Comment (lewtun) [inline recipes/dataset_filtering/config_demo.yaml:26]: Technically we filter on pass rate, so let's use that here:
```suggestion
pass_rate_min: 0.2
```

File: recipes/dataset_filtering/config_demo.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+high_reward_threshold: 0.8
```
Comment (lewtun) [inline recipes/dataset_filtering/config_demo.yaml:27]:
```suggestion
pass_rate_max: 0.8
```

File: recipes/dataset_filtering/config_demo.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+num_generations: 4
+temperature: 0.7
+
+# Reward func arguments
+reward_funcs:
+- accuracy
+reward_weights:
+- 1.0
+
+# Filtering arguments. Samples with mean reward outside of low / high will be filtered
```
Comment (lewtun) [inline recipes/dataset_filtering/config_demo.yaml:25]:
```suggestion
# Filtering arguments. Samples with a pass rate outside the interval `pass_rate_min < x < pass_rate_max` will be filtered.
```

**Review by lewtun — APPROVED** [review]
LGTM with some small suggestions to combine the subsets as configs and align the variables with pass rate

File: scripts/filter_dataset.py (hunk 5 above)
Comment (edbeeching) [reply scripts/filter_dataset.py:147]: I was thinking to change the dataset name to include the model and revision instead?

File: scripts/filter_dataset.py (hunk 2 above)
Comment (edbeeching) [reply scripts/filter_dataset.py:103]: I will expose it, but I think we should be using the same generation parameters here as we used for RL, rather than evaluation? As that is what we will use the samples for.
